src/components/utils/src/socket_win.cc

```diff
@@ -34,195 +34,257 @@
 }
 
-utils::Socket::Impl& utils::Socket::Impl::operator=(Impl& rh) {
+utils::TcpServerSocket::Impl::~Impl() {
   Close();
-  socket_ = rh.socket_;
-  rh.socket_ = NULL;
-  return *this;
 }
 
-bool utils::Socket::Impl::Valid() const {
-  return socket_ != NULL;
+bool utils::TcpServerSocket::Impl::IsListening() const {
+  return server_socket_ && is_listening_;
 }
 
-bool utils::Socket::Impl::Create(int af, int type, int protocol) {
-  if (!Close()) {
+bool utils::TcpServerSocket::Impl::Close() {
+  return CloseSocket(server_socket_);
+}
+
+bool utils::TcpServerSocket::Impl::Listen(
+    const std::string& address, int port, int backlog) {
+  LOG4CXX_AUTO_TRACE(logger_ptr);
+  if (IsListening()) {
+    LOG4CXX_ERROR(logger_ptr, "Cannot listen. Already listeneing.");
     return false;
   }
-  socket_ = socket(af, type, protocol);
-  if (INVALID_SOCKET == socket_) {
-    socket_ = NULL;
+
+  server_socket_ = socket(AF_INET, SOCK_STREAM, 0);
+  if (INVALID_SOCKET == server_socket_) {
+    LOG4CXX_ERROR(logger_ptr, "Failed to create server socket: " << WSAGetLastError());
+    server_socket_ = NULL;
     return false;
   }
-  return true;
-}
 
-bool utils::Socket::Impl::Close() {
-  if (NULL == socket_) {
-    return true;
-  }
-  if (-1 != closesocket(socket_)) {
+  char optval = 1;
+  if (SOCKET_ERROR == setsockopt(
+    server_socket_, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval))) {
+    LOG4CXX_ERROR(logger_ptr, "Unable to set sockopt: " << WSAGetLastError());
+    server_socket_ = NULL;
     return false;
   }
-  socket_ = NULL;
-  return true;
-}
 
-bool utils::Socket::Impl::SetOptions(
-  int level, int optname, const char* optval, size_t optlen) {
-  if (NULL == socket_ || SOCKET_ERROR == setsockopt(
-      socket_, level, optname, (const char*)&optval, optlen)) {
+  struct sockaddr_in server_address = { 0 };
+  server_address.sin_addr.s_addr = inet_addr(address.c_str());
+  server_address.sin_family = AF_INET;
+  server_address.sin_port = htons(port);
+
+  if (SOCKET_ERROR == bind(
+    server_socket_, reinterpret_cast<struct sockaddr*>(&server_address), sizeof(server_address))) {
+    LOG4CXX_ERROR(logger_ptr, "Unable to bind: " << WSAGetLastError());
+    server_socket_ = NULL;
     return false;
   }
-  return true;
-}
 
-bool utils::Socket::Impl::Bind(
-  const struct sockaddr* addr, size_t addrlen) {
-  if (NULL == socket_ || SOCKET_ERROR == bind(socket_, addr, addrlen)) {
+  if (SOCKET_ERROR == listen(server_socket_, backlog)) {
+    LOG4CXX_ERROR(logger_ptr, "Unable to listen: " << WSAGetLastError());
+    server_socket_ = NULL;
     return false;
   }
+
+  is_listening_ = true;
   return true;
 }
 
-bool utils::Socket::Impl::Listen(int backlog) {
-  if (NULL == socket_ || SOCKET_ERROR == listen(socket_, backlog)) {
-    return false;
+utils::TcpSocketConnection utils::TcpServerSocket::Impl::Accept() {
+  LOG4CXX_AUTO_TRACE(logger_ptr);
+
+  struct sockaddr client_addr = { 0 };
+  int client_addr_length = sizeof(client_addr);
```

Comment:
Seems you're wrong, client_addr_length  is in-out param. Thus is't not const. Look, it passed by pointer.